verifier/verifier.go

```diff
@@ -518,6 +540,78 @@ func verifyAuthenticTimestamp(outcome *notation.VerificationOutcome) *notation.V
 	}
 }
 
+func verifyRevocation(outcome *notation.VerificationOutcome, r revocation.Revocation, logger log.Logger) *notation.ValidationResult {
+	if r == nil {
+		return &notation.ValidationResult{
+			Type:   trustpolicy.TypeRevocation,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+			Error:  fmt.Errorf("unable to check revocation status, revocation client cannot be nil"),
+		}
+	}
+
+	signingTime := outcome.EnvelopeContent.SignerInfo.SignedAttributes.SigningTime
+	if signingTime.IsZero() {
+		signingTime = time.Now()
+	}
+	revocationResult, err := r.Validate(outcome.EnvelopeContent.SignerInfo.CertificateChain, signingTime)
+	if err != nil {
+		logger.Debug("error while checking revocation status, err: %s", err.Error())
+		return &notation.ValidationResult{
+			Type:   trustpolicy.TypeRevocation,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+			Error:  fmt.Errorf("unable to check revocation status, err: %s", err.Error()),
+		}
+	}
+
+	result := &notation.ValidationResult{
+		Type:   trustpolicy.TypeRevocation,
+		Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+	}
+
+	finalResult := revocation_result.ResultUnknown
+	numOKResults := 0
+	var problematicCertSubject string
+	revokedFound := false
+	var revokedCertSubject string
+	for i := len(revocationResult) - 1; i >= 0; i-- {
+		for _, serverResult := range revocationResult[i].ServerResults {
+			if serverResult != nil {
+				logger.Debugf("error for certificate #%d in chain with subject %v for server %q: %v", (i + 1), outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String(), serverResult.Server, serverResult.Error)
+			}
+		}
+
+		if revocationResult[i].Result == revocation_result.ResultOK || revocationResult[i].Result == revocation_result.ResultNonRevokable {
+			numOKResults++
+		} else {
+			finalResult = revocationResult[i].Result
+			problematicCertSubject = outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String()
+			if revocationResult[i].Result == revocation_result.ResultRevoked {
+				revokedFound = true
+				revokedCertSubject = problematicCertSubject
+			}
+		}
+	}
+	if revokedFound {
+		problematicCertSubject = revokedCertSubject
+		finalResult = revocation_result.ResultRevoked
+	}
```

Comment:
I'm thinking if we could make this part of code more elegant. Please correct me if I'm wrong, if the program could reach to line 589 `revokedFound = true`, looks like the program will always reach to line 595 `problematicCertSubject = revokedCertSubject`, because once `revokedFound` is set to true, it will stay to be true for the rest of the program. In this case, can we just set up `result.Error` and return `result` at the place where we set `revokedFound`, i.e., line 589?

Reply:
The issue with this is related to the subject. We want to use the leafmost revoked cert's subject. Thus, we are iterating backwards and do not break out (in case there is one closer to the leaf). This entire check for 594 is related to this subject for the error. This was the result of discussions about simplifying the logic with @rgnote and @priteshbandi

Reply:
Oh, I see. My initial understanding is "whenever we find a revoked cert in a cert chain, then the cert chain is immediately considered revoked, and we don't need to check the remaining certs in the chain anymore."
If we always want the leaf-most revoked cert's subject to be returned, then yes, the current logic is necessary.

Reply:
This was actually how I had it before. The issue with this is that the logic for differentiating between Unknown and Revoked becomes more complicated (more specifically, how to ensure that the leafmost cert subject for Unknown is used without being able to break). There is a trade-off for either approach, so the current one was decided as the most understandable to achieve both a precedence for Revoked over Unknown and using the subject for the leafmost cert for the appropriate result.